### PR TITLE
Fix parser error for using statements

### LINKED ISSUE #17113: PSReadLine throws an exception when typing a type name after `using namespace [NameSpace]`
### Exception report

```console
### Environment
PSReadLine: 2.2.2
PowerShell: 7.3.0-preview.2
OS: Microsoft Windows 10.0.19044
BufferWidth: 208
BufferHeight: 46

Last 5 Keys:

 UpArrow UpArrow Enter
 [ e

### Exception

System.Management.Automation.PSArgumentException: Cannot process argument because the value of argument "name" is not valid. Change the value of the "name" argument and run the operation again.
   at System.Management.Automation.Language.TypeName..ctor(IScriptExtent extent, String name)
   at System.Management.Automation.Language.TypeResolver.ResolveTypeNameWithContext(TypeName typeName, Exception& exception, Assembly[] assemblies, TypeResolutionState typeResolutionState)
   at System.Management.Automation.Language.SymbolResolver.VisitTypeName(TypeName typeName, Int32 genericArgumentCount, Boolean isAttribute)
   at System.Management.Automation.Language.SymbolResolver.DispatchTypeName(ITypeName type, Int32 genericArgumentCount, Boolean isAttribute)
   at System.Management.Automation.Language.SymbolResolver.VisitTypeExpression(TypeExpressionAst typeExpressionAst)
   at System.Management.Automation.Language.TypeExpressionAst.InternalVisit(AstVisitor visitor)
   at System.Management.Automation.Language.CommandExpressionAst.InternalVisit(AstVisitor visitor)
   at System.Management.Automation.Language.PipelineAst.InternalVisit(AstVisitor visitor)
   at System.Management.Automation.Language.StatementBlockAst.InternalVisit(AstVisitor visitor, ReadOnlyCollection`1 traps, ReadOnlyCollection`1 statements, AstVisitAction action)
   at System.Management.Automation.Language.NamedBlockAst.InternalVisit(AstVisitor visitor)
   at System.Management.Automation.Language.ScriptBlockAst.InternalVisit(AstVisitor visitor)
   at System.Management.Automation.Language.Ast.Visit(AstVisitor astVisitor)
   at System.Management.Automation.Language.SymbolResolver.ResolveSymbols(Parser parser, ScriptBlockAst scriptBlockAst)
   at System.Management.Automation.Language.ScriptBlockAst.PerformPostParseChecks(Parser parser)
   at System.Management.Automation.Language.Parser.ParseTask(String fileName, String input, List`1 tokenList, Boolean recursed, ParseMode parseMode)
   at System.Management.Automation.Language.Parser.Parse(String fileName, String input, List`1 tokenList, ParseError[]& errors, ParseMode parseMode)
   at System.Management.Automation.Language.Parser.ParseInput(String input, String fileName, Token[]& tokens, ParseError[]& errors)
```

### Screenshot

image

### Steps to reproduce

```
$> using namespace [NameSpace]
$> [A ...
```
The namespace doesn't have to exist, just make up anything. The important part are the brackets around the namespace – they shouldn't be there, but apparently it is syntactically valid (not sure about semantics) and causes PSReadLine to crash.

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
Fixes a parser error where an invalid namespace would cause the parser to crash when attempting to resolve a type name. For example: `using namespace [System];[S`
Also made some adjustments in the completion code so namespace completion continues to work for invalid namespaces.

## PR Context
Fixes https://github.com/PowerShell/PowerShell/issues/17113

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by iSazonov** [comment]
What about `using namespace !!!.!!!`?

**Comment by MartinGC94** [comment]
> What about `using namespace !!!.!!!`?

I haven't checked, but I'm fairly certain it's still allowed with this change. While it would certainly be nice if the parser prevented every invalid namespace entry, the main purpose of this change is just to fix the exception thrown when specific characters are used, see: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/ast.cs#L8457

**Comment by iSazonov** [comment]
> For this issue, there are 2 potential fixes:

> - make using namespace to error on invalid namespace names
> - make the constructor of TypeName to tolerate invalid namespace names
It feels to me option 1 is the right one to go with, but it's not trivial to validate if the namespace name is legitimate using the C# identifier definition.

Although @daxian-dbw pointed out that the first solution is preferable I think full solution should include both since it is impossible to manually check whether a string valid namespace without direct test with the constructor.
I'd prefer to do formal notation check on first step (with Regex) and then try constructor. This allows us to make useful error messages and suggestions and block incorrect inputs.

**Comment by MartinGC94** [comment]
@iSazonov Using statements cannot be so strict during parsing because types can be created on the fly like this:

```
using namespace DemoNamespace
Add-Type -TypeDefinition @'
namespace DemoNamespace
{
    public class Hello
    {
        public static void World()
        {
            System.Console.WriteLine("Hello World");
        }
    }
}
'@
[Hello]::World()
```
I agree that the syntax check itself could be made more thorough.

**Comment by iSazonov** [comment]
@MartinGC94 I mean the checks are performed at different times.
As for Regex we could start with simple one. I don't think we should be so strong as C# compiler. Our intention is to make useful error message instead of unexpected exception.

**Comment by MartinGC94** [comment]
@iSazonov If the intent is just to prevent unintentional exceptions then the current checks should be fine. Invalid namespaces like `using namespace !!!.!!!` won't throw any errors but they obviously won't do anything useful either.  
If you want, I can easily write a regex with the rules I think namespace identifiers have, I'm just saying there's a good chance I'll make a mistake because I don't fully understand the spec.

**Comment by iSazonov** [comment]
You could play with `[regex]::IsMatch("a.q.w", "^([\p{L}_][\p{L}\p{Pc}\p{Nd}\p{Cf}\p{M}]*)(\.[\p{L}_][\p{L}\p{Pc}\p{Nd}\p{Cf}\p{M}]*)*$")`

For reference https://www.regular-expressions.info/unicode.html

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 1]
```diff
@@ -4945,7 +4945,8 @@ private StatementAst UsingStatementRule(Token usingToken)
                 return new ErrorStatementAst(ExtentOf(usingToken, itemToken.Extent));
             }
 
-            if (itemAst is StringConstantExpressionAst stringAst
+            if (kind is UsingStatementKind.Namespace or UsingStatementKind.Type
+                && itemAst is StringConstantExpressionAst stringAst
                 && !Regex.IsMatch(stringAst.Value, @"^([\p{L}_][\p{L}\p{Pc}\p{Nd}\p{Cf}\p{M}]*)(\.[\p{L}_][\p{L}\p{Pc}\p{Nd}\p{Cf}\p{M}]*)*$"))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs:4950]: Please add comment with daxian-dbw's links to explain the Regex.

**Comment by JamesWTruher** [comment]
while I agree that this is a problem, I'm concerned about this approach. It seems that the issue is that `new TypeName` is not expected to throw and it clearly will under some circumstances. A simpler approach might be to address _that_.
I tried this approach and it seems to produce the proper result:

```csharp
TypeName newTypeName;
try
{
   newTypeName = new TypeName(typeName.Extent, newTypeNameToSearch);
}
catch (Exception e)
{
   exception = e;
   break;
}
```

this is around line 400 in engine/parser/TypeResolver.cs
here's the error message with this implementation:

```powershell
PS> using namespace [system];[s                   
ParserError: 
Line |
   1 |  using namespace [system];[s
     |                             ~
     | Missing ] at end of attribute or type literal.
```

After more consideration, a `TypeName.TryCreate` might be an even better approach (to enable better inlining) then checking to see if we were able to create it.

**Review by JamesWTruher — CHANGES_REQUESTED** [review]
please see my comments in the conversation

**Comment by MartinGC94** [comment]
@JamesWTruher I agree that adding `TypeName.TryCreate` could be useful but should it be in place of the existing changes that makes the namespace parsing more strict? Or in addition to it? I think it would be useful to have the parser show an error for invalid namespaces.

**Comment by daxian-dbw** [comment]
Superseded by #21162. @MartinGC94 and @iSazonov Thanks for your initial investigation! I have called out the thanks to both of you in the new PR.

**Pull request closed without merging**